File: 1-100/99. Recover Binary Search Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */
var recoverTree = function(root) {
    
    var prev, first, second
    const inode = (node) => {
        if(!node){
            return 
        }
        inode(node.left)
        if(prev && prev.val > node.val){
            if(!first){
                first = prev
            }
            if(first){
                second = node
            }
        }
        prev = node

        inode(node.right)
    }
    inode(root)
    const temp = first.val
    first.val = second.val
    second.val = temp
};

File: 1-100/99. Recover Binary Search Tree.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func recoverTree(root *TreeNode) {
	var first, prev, second *TreeNode
	var dfs func(root *TreeNode)
	dfs = func (root *TreeNode) {
		if root == nil {
			return
		}
		dfs(root.Left)
		if prev != nil && prev.Val > root.Val {
			if first == nil {
				first = prev
			} 
			if first != nil {
				second = root
			}
		}
		prev = root
		// do something with root.val
		dfs(root.Right)
	}
	dfs(root)
	first.Val, second.Val = second.Val, first.Val
}
